File: snalltaget.php
<?php
if(isset($_GET['debugg']))
	{
	error_reporting(E_ALL);
	ini_set('display_errors', 'on');
	}

if (isset($_GET['from']) == FALSE)
	{
	die('{"error":"data from missing"}');
	}

if (isset($_GET['to']) == FALSE)
	{
	die('{"error":"data to missing"}');
	}

if (isset($_GET['date']) == FALSE)
	{
	die('{"error":"data date missing"}');
	}

if (isset($_GET['departureTime']) == FALSE)
	{
	die('{"error":"data from missing"}');
	}

$ch = curl_init();
$temp = tmpfile();

curl_setopt( $ch, CURLOPT_RETURNTRANSFER, true );
curl_setopt($ch, CURLOPT_URL, 'https://boka.snalltaget.se/boka-biljett');
curl_setopt($ch, CURLOPT_HTTPHEADER, array(
'Accept-Encoding: gzip,deflate,sdch' 
, 'Host: boka.snalltaget.se' 
, 'Accept-Language: sv-SE,sv;q=0.8,en-US;q=0.6,en;q=0.4'
, 'User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/30.0.1599.114 Chrome/30.0.1599.114 Safari/537.36' 
, 'Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8' 
, 'Cache-Control: max-age=0'));
curl_setopt( $ch, CURLOPT_COOKIEJAR,  $temp );
curl_setopt( $ch, CURLOPT_COOKIEFILE, $temp );



if(file_exists('snalltagetcache/snalltaget.json') === TRUE)
	
	{
	$allstops = json_decode(file_get_contents('snalltagetcache/snalltaget.json'));
	}

else	
	{
	curl_exec($ch);
	curl_setopt($ch, CURLOPT_URL, 'https://boka.snalltaget.se/api/locations/all');
	$data = curl_exec($ch);
	$allstops = json_decode($data);
	file_put_contents('snalltagetcache/snalltaget.json',json_encode($allstops));
	}

$stoparray = array();
foreach($allstops->frequent as $stop)
	{
	$stoparray[strval(intval($stop->P)*100000+intval($stop->L))] = $stop;
	}

foreach($allstops->infrequent as $stop)
	{
	$stoparray[strval(intval($stop->P)*100000+intval($stop->L))] = $stop;
	}

if(isset($stoparray[$_GET['to']]) AND isset($stoparray[$_GET['from']]))
	{		
	curl_exec($ch);
	}
else	{
	die('{"error":"stop not in network"}');
	}

curl_setopt($ch, CURLOPT_URL, 'https://boka.snalltaget.se/api/timetables');
curl_setopt($ch, CURLOPT_HTTPHEADER, array(
'Origin: https://boka.snalltaget.se'
,'Accept-Encoding: gzip,deflate,sdch' 
,'Host: boka.snalltaget.se'
,'Accept-Language: sv-SE' 
,'User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/30.0.1599.114 Chrome/30.0.1599.114 Safari/537.36' 
,'Content-Type: application/json;charset=UTF-8' 
,'Accept: application/json, text/plain, */*' 
,'Referer: https://boka.snalltaget.se/boka-biljett' 
,'Connection: keep-alive'));

$request = '{"DepartureLocationId":'.intval(substr($_GET['from'],2)).
',"DepartureLocationProducerCode":'.intval(substr($_GET['from'],0,2)).
',"ArrivalLocationId":'.intval(substr($_GET['to'],2)).
',"ArrivalLocationProducerCode":'.intval(substr($_GET['to'],0,2)).
',"DepartureDateTime":"'.$_GET['date'].' '.$_GET['departureTime'].'"'.
',"TravelType":"E","Passengers":[{"PassengerCategory":"VU"}]}';

curl_setopt($ch, CURLOPT_POSTFIELDS,$request);
curl_setopt($ch, CURLOPT_POST, TRUE);
$triplist = json_decode(curl_exec($ch));

$list = "";
$comma = "";
foreach($triplist->JourneyAdvices as $refid){
$list = $list.$comma.$refid->JourneyConnectionReference;
$comma = ",";
}

curl_setopt($ch, CURLOPT_URL, 'https://boka.snalltaget.se/api/journeyadvices/lowestprices');
curl_setopt($ch, CURLOPT_POSTFIELDS, '{"TimetableId":"'.$triplist->Id.'","JourneyConnectionReferences":['.$list.']}');
$price = json_decode(curl_exec($ch));

$findprice = array();

foreach($price as $row){
$findprice[$row->JourneyConnectionReference]=$row;
}

foreach($triplist->JourneyAdvices as $key => $refid){
$triplist->JourneyAdvices[$key]->PriceData = $findprice[$refid->JourneyConnectionReference];
}

fclose($temp);

print json_encode($triplist);

// 'https://boka.snalltaget.se/boka-biljett#!/step1?from=1&to=120&date=2013-12-15';

?>

File: snalltaget.py
import requests
import re
import json
import tornado.ioloop
import tornado.web
from threading import Thread

cache = {}

class MainHandler(tornado.web.RequestHandler):
	@tornado.web.asynchronous
	def get(self):
		Thread(target=self.makerequest).start()
		
	def returnrequest(self, data):
		tornado.ioloop.IOLoop.instance().add_callback(self.returndata, data)
	
	def makerequest(self):
		global cache
		r = requests.get('https://boka.snalltaget.se/boka-biljett')
		cookie = r.cookies["Token"]
		cookies = dict(Token=cookie)

		query = json.loads('{"DepartureLocationId":1,"DepartureLocationProducerCode":74,"ArrivalLocationId":110,"ArrivalLocationProducerCode":74,"DepartureDateTime":"2014-02-21 12:00","TravelType":"E","Passengers":[{"PassengerCategory":"VU"}]}')
		#query['DepartureDateTime'] = self.get_argument('date','2014-02-20') + self.get_argument('departureTime','12:00')
		headers = {'content-type': 'application/json'}

		r = requests.post('https://boka.snalltaget.se/api/timetables', data=json.dumps(query), headers=headers, cookies=cookies)

		trips = r.json()
		
		if len(trips['JourneyAdvices']) > 10:
			max = 10
		else:
			max = len(trips['JourneyAdvices']-1)
			
		pquery = {'TimetableId':trips['Id'], 'JourneyConnectionReferences':[]}

		for i in range(0, max):
			pquery['JourneyConnectionReferences'].append(trips['JourneyAdvices'][i]['JourneyConnectionReference'])

		r = requests.post('https://boka.snalltaget.se/api/journeyadvices/lowestprices', data=json.dumps(pquery), headers=headers, cookies=cookies)

		price = r.json()

		for i in range(0, len(trips['JourneyAdvices'])):
			for j in range (0,len(price)):
				if price[j]['JourneyConnectionReference'] == trips['JourneyAdvices'][i]['JourneyConnectionReference']:
					trips['JourneyAdvices'][i]['IsSleeperTrain'] = price[j]['IsSleeperTrain']
					trips['JourneyAdvices'][i]['LowestTotalPrice'] = price[j]['LowestTotalPrice']
					trips['JourneyAdvices'][i]['Currency'] = price[j]['Currency']
					stopfrom = str(trips['JourneyAdvices'][i]['DepartureLocation']['ProducerCode']*100000 + trips['JourneyAdvices'][i]['DepartureLocation']['LocationId'])
					stopto = str(trips['JourneyAdvices'][i]['ArrivalLocation']['ProducerCode']*100000 +trips['JourneyAdvices'][i]['ArrivalLocation']['LocationId'])
					timefrom = trips['JourneyAdvices'][i]['DepartureDateTime']
					timeto = trips['JourneyAdvices'][i]['ArrivalDateTime']
					cache[stopfrom+timefrom+stopto+timeto] = trips['JourneyAdvices'][i]
					break
		print cache
		self.returnrequest(trips)
		
	def returndata(self, trips):
		self.write(trips)
		self.finish()

application = tornado.web.Application([
    (r"/snalltaget/", MainHandler),
])

application.listen(8888)
tornado.ioloop.IOLoop.instance().start()